controllers/cloud.redhat.com/clowdenvironment_reconciliation.go

```diff
@@ -0,0 +1,489 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	crd "github.com/RedHatInsights/clowder/apis/cloud.redhat.com/v1alpha1"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/clowderconfig"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/errors"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/providers"
+	rc "github.com/RedHatInsights/rhc-osdk-utils/resource_cache"
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	cond "sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	SKIPRECONCILE = "SKIPRECONCILE"
+)
+
+//During reconciliation we handle errors in 2 ways: sometimes we want to error out of reconciliation and sometimes we want to skip reconciliation.
+func shouldSkipReconciliation(err error) bool {
+	if err != nil && err.Error() == SKIPRECONCILE {
+		return true
+	}
+	return false
+}
+
+func MakeClowdEnvReconciliation(cache *rc.ObjectCache,
+	recorder record.EventRecorder,
+	ctx *context.Context,
+	client client.Client,
+	env *crd.ClowdEnvironment,
+	log *logr.Logger) ClowdEnvironmentReconciliation {
+	return ClowdEnvironmentReconciliation{
+		cache:    cache,
+		recorder: recorder,
+		ctx:      ctx,
+		client:   client,
+		env:      env,
+		log:      log,
+	}
+}
+
+//ClowdEnvironmentReconciliation represents a single reconciliation event
+//This type is created by ClowdEnvironmentReconciler which handles the reconciliation cycle as a whole
+//ClowdEnvironmentReconciliation encapsulates all of the state and logic requires for a single
+//reconciliation event
+type ClowdEnvironmentReconciliation struct {
+	cache    *rc.ObjectCache
+	recorder record.EventRecorder
+	ctx      *context.Context
+	client   client.Client
+	env      *crd.ClowdEnvironment
+	log      *logr.Logger
+}
+
+//Returns a list of step methods that should be run during reconciliation
+func (r *ClowdEnvironmentReconciliation) steps() []func() (ctrl.Result, error) {
```

Comment:
In thinking about the steps piece here - the only thing I'm concerned about is that if we need to inject something in the middle it's a whole new step - I guess that's OK, yeh I guess that's fine - sorry stream of consciousness

Reply:
Yeah I would actually say that's a huge benefit. The reconcile function before was very hard to understand for someone new (or even not so knew to the project) - really had to take it in like a novel and understand it cover to cover. You had to read it and say "OK we're doing this here, and doing this here" etc. Now you can just read the step list and its like "Oh, ok, I get that." And if you want to add a step you know exactly where you are in the process, what's come before and will happen after, etc. Its maybe my favorite piece of this refactor. Seeing reconciliation as this list of steps I think is a bonus.